### PR TITLE
Add moving mesh capabilities to boundary coordinates tag

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Combine test cases
- Add `ASSERT`s to get useful error messages when an expected function of time isn't available
- Make the `Tags::BoundaryCoordinates` interface compute tag work with a moving mesh

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Domain/InterfaceComputeTags.hpp [hunk 1]
```diff
@@ -11,17 +11,24 @@
...
 #include "DataStructures/Tensor/Slice.hpp"
+#include "Domain/CoordinateMaps/Tags.hpp"
 #include "Domain/Direction.hpp"
 #include "Domain/ElementMap.hpp"
 #include "Domain/IndexToSliceAt.hpp"
 #include "Domain/InterfaceHelpers.hpp"
 #include "Domain/LogicalCoordinates.hpp"
 #include "Domain/Mesh.hpp"
 #include "Domain/Tags.hpp"
+// #include "Domain/TagsTimeDependent.hpp"
```
Comment (kidder) [inline src/Domain/InterfaceComputeTags.hpp]: ?

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

**Comment by nilsdeppe** [comment]
rebased and squashed

**Review by kidder — DISMISSED** [review]

File: tests/Unit/Domain/Test_InterfaceItems.cpp [hunk 2]
```diff
@@ -475,6 +486,172 @@ void test_interface_slice(){
...
+        make_coordinate_map_base<Frame::Grid, Frame::Inertial>(
+            CoordMapsTimeDependent::CubicScale<1>{outer_boundary,
+                                                  functions_of_time_names[0],
+                                                  functions_of_time_names[1]});
+    const auto logical_to_inertial_map =
+        make_coordinate_map_base<Frame::Logical, Frame::Inertial>(
+            CoordinateMaps::Affine(-1.0, 1.0, 2.0, 7.8),
+            CoordMapsTimeDependent::CubicScale<1>{outer_boundary,
+                                                  functions_of_time_names[0],
+                                                  functions_of_time_names[1]});
```
Comment (fmahebert) [inline tests/Unit/Domain/Test_InterfaceItems.cpp]: Here (and in 2d, 3d cases below), do you think it would be reasonable to structure things like this?
```cpp
const ... logical_to_grid_func = CoordinateMaps::Affine(-1.0, 1.0, 2.0, 7.8);
const ... grid_to_inertial_func = CoordMapsTimeDependent::CubicScale<1>{ ... };
const ElementMap<1, Frame::Grid> logical_to_grid_map(..., make_coordinate_map_base<...>(logical_to_grid_func);
const auto grid_to_inertial_map = make_coordinate_map_base<...>(grid_to_inertial_func);
const auto logical_to_inertial_map = make_coordinate_map_base<...>(logical_to_grid_func, grid_to_inertial_func);
```

The goal would be to avoid having to specify each coord map two separate times... But this does add more lines of code, so it's up to you whichever you think is the cleanest approach 🙂

File: tests/Unit/Domain/Test_InterfaceItems.cpp [hunk 3]
```diff
@@ -520,5 +697,6 @@ SPECTRE_TEST_CASE("Unit.Domain.InterfaceItems", "[Unit][Domain]") {
   test_interface_subitems();
   test_interface_slice();
   test_interface_base_tags();
+  test_boundary_coordinates_moving_mesh();
```
Comment (fmahebert) [inline tests/Unit/Domain/Test_InterfaceItems.cpp:664]: [nitpick] Make order of test function definitions consistent with order of test function calls... either by moving this up one line, or by moving the new test functions down under `test_interface_base_tags`.

**Review by fmahebert — DISMISSED** [review]
Overall this looks good to me, but here are two non-binding thoughts. If you do make any changes, please squash them directly.

**Comment by nilsdeppe** [comment]
Rebased on develop, made the suggested changes and squashed. Thanks for the reviews :)

File: tests/Unit/Domain/Test_InterfaceItems.cpp (hunk 2 above)
Comment (fmahebert) [reply tests/Unit/Domain/Test_InterfaceItems.cpp]: @nilsdeppe as a sanity check ... I see you updated the 2d and 3d map tests per this suggestion, but not the 1d test. Was this intentional? Or did you miss the 1d case?

Comment (nilsdeppe) [reply tests/Unit/Domain/Test_InterfaceItems.cpp]: Nah, I'm just dumb :P

**Comment by nilsdeppe** [comment]
Okay, added a new commit `Add volume_dim member variable to ElementId` and did some more refactor that further reduces code duplication in the test :)

**Review by fmahebert — APPROVED** [review]

File: src/Domain/ElementId.cpp [hunk 4]
```diff
@@ -15,6 +15,9 @@
 #include "Utilities/MakeArray.hpp"
 #include "Utilities/StdHelpers.hpp"  // IWYU pragma: keep
 
+template <size_t VolumeDim>
```
Comment (kidder) [inline src/Domain/ElementId.cpp:18]: why is this needed?

Comment (nilsdeppe) [reply src/Domain/ElementId.cpp:18]: This is the definition of the member variable and allows it to be linked to from other libraries. C++ doesn't implicitly define `static constexpr` member variables

**Review by kidder — APPROVED** [review]
